### ISSUE TITLE
Dart 2 Compatability

### ISSUE THREAD

**Issue by chances** [agilord/owl#4]
I'm using this library indirectly with [spotify-dart](https://github.com/rinukkusu/spotify-dart) in a [Flutter](https://flutter.io/) app that is AOT compiled with Dart v2.0.0-dev.67.0, but compilation fails because of Dart 2's strict types due to the `dynamic` type parameter generated in [`json_generator.dart:81`](https://github.com/agilord/owl/blob/2c27015342b42b6448310fa0d23df66868982e19/owl_codegen/lib/json_generator.dart#L81).

i.e. `(map['${field.keyName}'] as List<dynamic>)?.map(${field.parserFn})?.toList()`

Flutter compiler error:
```
compiler message: file:///Users/chancesnow/bin/flutter/.pub-cache/git/spotify-dart-55bff18594c7337681570999f994bdf8cf42b0ba/lib/spotify_io.json.g.dart:129:59: Error: The argument type '(dart.core::Map<dart.core::String, dynamic>) → spotify::Image' can't be assigned to the parameter type '(dynamic) → dynamic'.
compiler message: Try changing the type of the parameter, or casting the argument to '(dynamic) → dynamic'.
compiler message:         (map['images'] as List<dynamic>)?.map(ImageMapper.parse)?.toList();
```

I've tried changing `List<dynamic>` to `List<Map<String, dynamic>>` in my fork of Owl to no avail:

<details><summary>pub run owl_codegen:main -t json</summary>
<p>

```
Unable to spawn isolate: file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/typed/stream.dart:56:10: Error: The return type of the method 'TypeSafeStream::firstWhere' is dart.async::Future<dynamic>, which does not match the return type of the overridden method (dart.async::Future<#lib1::TypeSafeStream::T>).
Change to a subtype of dart.async::Future<#lib1::TypeSafeStream::T>.
  Future firstWhere(bool test(T element), {Object defaultValue()}) =>
         ^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/typed/stream.dart:56:10: Error: The method 'TypeSafeStream::firstWhere' doesn't have the named parameter 'orElse' of overriden method 'Stream::firstWhere'.
  Future firstWhere(bool test(T element), {Object defaultValue()}) =>
         ^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/typed/stream.dart:59:10: Error: The return type of the method 'TypeSafeStream::lastWhere' is dart.async::Future<dynamic>, which does not match the return type of the overridden method (dart.async::Future<#lib1::TypeSafeStream::T>).
Change to a subtype of dart.async::Future<#lib1::TypeSafeStream::T>.
  Future lastWhere(bool test(T element), {Object defaultValue()}) =>
         ^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/typed/stream.dart:59:10: Error: The method 'TypeSafeStream::lastWhere' doesn't have the named parameter 'orElse' of overriden method 'Stream::lastWhere'.
  Future lastWhere(bool test(T element), {Object defaultValue()}) =>
         ^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/typed/stream.dart:62:13: Error: The method 'TypeSafeStream::singleWhere' has fewer named arguments than those of overridden method 'Stream::singleWhere'.
  Future<T> singleWhere(bool test(T element)) async =>
            ^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/build_runner-0.7.13+1/lib/src/performance_tracking/performance_tracking_resolvers.dart:11:7: Error: The non-abstract class 'PerformanceTrackingResolvers' is missing implementations for these members:
  'reset'.
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class PerformanceTrackingResolvers implements Resolvers {
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/restartable_timer.dart:11:7: Error: The non-abstract class 'RestartableTimer' is missing implementations for these members:
  'tick'.
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class RestartableTimer implements Timer {
      ^^^^^^^^^^^^^^^^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/result/capture_transformer.dart:15:7: Error: The non-abstract class 'CaptureStreamTransformer' is missing implementations for these members:
  'cast'.
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class CaptureStreamTransformer<T> implements StreamTransformer<T, Result<T>> {
      ^^^^^^^^^^^^^^^^^^^^^^^^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/result/release_transformer.dart:12:7: Error: The non-abstract class 'ReleaseStreamTransformer' is missing implementations for these members:
  'cast'.
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.

class ReleaseStreamTransformer<T> implements StreamTransformer<Result<T>, T> {
      ^^^^^^^^^^^^^^^^^^^^^^^^
file:///Users/chancesnow/.pub-cache/hosted/pub.dartlang.org/async-1.13.3/lib/src/single_subscription_transformer.dart:16:7: Error: The non-abstract class 'SingleSubscriptionTransformer' is missing implementations for these members:
  'cast'.
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.
```

</p>
</details>
<p>

I've tried upgrading `analyzer`, `build_runner`, and `source_gen` to their latest versions for the `owl_codegen` package to no avail. (Same errors as above.)

I'm not sure what else to try. Any ideas?

**Comment by olmedocr** [comment]
Pretty interested in fix this too, Is there some documentation to see how owl_codegen works?

**Comment by isoos** [comment]
I've released a new version that should hopefully work with Dart 2. I haven't got much chance to test it yet, because the codebase that I'm using it in is still on Dart 1. Let me know if it works.

Btw. I'm using a new, much simpler codegen for my database access, using the `postgres` package (which is at least maintained). Let me know if there is an interest to publish that too.

**Comment by rinukkusu** [comment]
I just ran into this when trying to use owl 0.2.3 with Dart SDK 2.0:

```
Because owl >=0.2.3 depends on postgresql >=0.2.1 which requires SDK version <2.0.0, owl >=0.2.3 is forbidden.
So, because darksky_weather depends on owl ^0.2.3, version solving failed.
```

**Comment by isoos** [comment]
Yeah, the `postgresql` library is abandoned, and in my project I'm using `postgres` for a while now. I have a new kind of SQL generator for that, and I see no way to have an upgrade without a breaking change.

Would anybody share their use-case of `owl`?

**Comment by rinukkusu** [comment]
Well, I've just used it in a few libraries for JSON deserialisation, which I wanted to upgrade to Dart SDK 2.0.

**Comment by isoos** [comment]
I think for the JSON use-case [json_serializable](https://pub.dartlang.org/packages/json_serializable) provides a slightly different, but probably good-enough replacement. There are some differences, e.g. it relies on `dart:convert` to fully expand (and serialize) the object graph to maps and lists and primitives, while `owl` did that on its own.

For the sql support I have now a different code generator for `package:postgres`, and I'm planning to release it in a few weeks (pending on how much time I have). I believe the web part was not really used.

**Comment by isoos** [comment]
@rinukkusu wdyt about `json_serializable`?

**Comment by isoos** [comment]
I've double-checked my codebase, and I actually have a code where I need to upgrade owl to Dart 2 before I could migrate the rest of it. I'll start working on it soon.

**Comment by rinukkusu** [comment]
I've already successfully migrated a few of my private packages to `json_serializable` over the past week, so don't worry or hurry :smile: (I can't push to Pub atm anyway lol)

**Comment by isoos** [comment]
Just published `0.4.0`. It has a breaking change for SQL, but I couldn't resolve it any other way. I'll be testing it in more depth tomorrow.
